### PR TITLE
Fix polyphonic audio streams with id > 1 cannot be stopped or changed (MSVC mis-optimization)

### PR DESCRIPTION
**Pull request by ditiem-games**
Fixes https://github.com/godotengine/godot/issues/86053

### CODE REVIEW

**Comment by Chaosus** [comment]
This is not an acceptable name for the commit/PR. Please rename it to meaningful one (like "Fix audio streams with id > 1 cannot be stopped or changed") and squash the commits.

**Review by bruvzg — COMMENTED** [review]
If this logic is removed, then these two lines in the header should be removed as well:

https://github.com/godotengine/godot/blob/475248d99df89fc29032a42f1d29ad4cef49c8b5/scene/resources/audio_stream_polyphonic.h#L59-L60

Why was it saving/using index in the first place? It might be performance critical, and in this case it's better to keep it, check index first and then fallback to the full list search instead of removing it completely.

**Comment by ditiem-games** [comment]
> If this logic is removed, then these two lines in the header should be removed as well:
Agree.

> Why was it saving/using index in the first place?

I thought about that for quite a while and my only conclusions was that whoever did it consider some use cases but forgot the delete song case or maybe in the moment it was implemented that case was not possible.

> It might be performance critical, and in this case it's better to keep it, check index first and then fallback to the full list search instead of removing it completely.

We need serious data for this, and I personally do not have it. I thought of the cases about how I did/will use it in my games and how it will be impacted. So my assumptions were:
1. There are not gonna be many songs, around 5 is the max I got (32 is the max allowed if I am not mistaken).
2. It is not very usual to call this function (just when stopping, updating the pitch/volume by some fade-out/in, for instance).
3. By the moment that you end with the checks of whether the "id" was correct and you can "optimise" the search, you are probably in the 3rd of 4th song using the for implementation. But for this affirmation, deep testing is really needed, and I do not know how to do that in the current project. If this point is true, accessing, in my cases, will take in average double the time.

If you want to reduce the linear access time, a hash function or AVL tree (or similar) should be used instead.

Is there any data about number of streams and how often are they accessed given the id?

**Comment by ditiem-games** [comment]
> Why was it saving/using index in the first place?

There is a second approach that can be taken which I think it is interesting. The streams vector:

https://github.com/godotengine/godot/blob/475248d99df89fc29032a42f1d29ad4cef49c8b5/scene/resources/audio_stream_polyphonic.h#L78

contains elements of type Stream. Here the rules:
1. It should be possible to create some "Null Stream" element that will substitute a deleted element, instead of collapsing the vector.
2. Whenever a new Stream is added, the "Null Stream" elements are checked first to use that index in the ID
3. The current implementation of index + stream-id will remain as it is.
4. Vector may shrink if all the end elements are "Null Stream", but I am not sure it is worth the shrinking.

Creation will be linear, but all the rest will be constant access. 

I will be away for a week and I will give it a try on my return.

**Comment by ditiem-games** [comment]
The idea in the previous comment is exactly what the author had in mind. So the bug was much deeper. 

The `ID_MASK` was defined as 0xFFFFFFFF, which was treated as an int so it was extended to 0xFFFFFFFFFFFFFFFF, failing to extract the low 32 bits of a uint64_t.

I would have personally used an struct with two uint32_t and avoid all these masks.

There was another issue that was not addressed. When the stream ended, it was not freed until another stream will take its slot, if that ever happened. I do not know if this was intended. I refactored the code to `unref` the stream.

**Comment by ditiem-games** [comment]
@bruvzg @Chaosus guys, can you have a look at this PR? It is ready for merge, the fix was way simpler at the end and nothing gets changed or broken.

**Comment by akien-mga** [comment]
See also #86054 for an alternative approach.

**Comment by akien-mga** [comment]
I slightly prefer #86054 as it's more minimal and prevents some extra casting (it seems weird to have a `int64_t` id but then always cast it internally to `uint64_t` before casting back to `int64_t`), but both solutions were valid. Thanks for the contribution!

**Comment by akien-mga** [comment]
> The enum is exported in the .h file unnecessarily because no other file uses it.

That's a common practice in the Godot codebase to put helper constants like this in an enum. But indeed it does lead to issues like this one with MSVC being peculiar with enums, so it doesn't hurt to refactor it on a case by case basis.

> The ID was already declared as an int. I tried changing it to uint, but it breaks some APIs. So, I went with casting as the only alternative left. It should be removed by the compiler in most cases, but if that's not the case and things are promoted to int, the bug remains.

Yeah there's no `uint` in the scripting API, `int` in GDScript is signed 64-bit. So I guess casting like you did does make sense.

Let's go with this approach then :)

File: scene/resources/audio_stream_polyphonic.cpp [hunk 1]
```diff
@@ -34,6 +34,9 @@
 #include "scene/main/scene_tree.h"
 #include "servers/audio_server.h"
 
+constexpr uint64_t ID_MASK = 0x0FFFFFFFF;
```
Comment (akien-mga) [inline scene/resources/audio_stream_polyphonic.cpp:37]: I first thought you had changed the mask with that extra leading 0, as I thought it was `0x0FFF_FFFF` before couting the `F`s.

Makes more sense to me to only include 8 hex and not 9, so our brains can safely assume that a long-ish `0xFF...FFF` might indeed be the common `0xFFFF_FFFF`.
```suggestion
constexpr uint64_t ID_MASK = 0xFFFF_FFFF;
```

Comment (akien-mga) [reply scene/resources/audio_stream_polyphonic.cpp]: Amended myself.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by Zireael07** [comment]
It's probably worth documenting in the code itself why it's done this way

**Pull request merged**